Версия 2.1

Akmula opened the pull request:
1. Создать интерфейсы FilmStorage и UserStorage.
2. Создать классы InMemoryFilmStorage и InMemoryUserStorage, имплементирующие новые интерфейсы, и перенести туда всю логику хранения, обновления и поиска объектов.
3. Добавить к InMemoryFilmStorage и InMemoryUserStorage аннотацию @Component, чтобы впоследствии пользоваться внедрением зависимостей и передавать хранилища сервисам.
4. Создать UserService, который будет отвечать за такие операции с пользователями, как добавление в друзья, удаление из друзей, вывод списка общих друзей.
5. Создать FilmService, который будет отвечать за операции с фильмами — добавление и удаление лайка, вывод 10 наиболее популярных фильмов по количеству лайков.
6. Переделатье код в контроллерах, сервисах и хранилищах под использование внедрения зависимостей.
7. Использовать аннотации @Service, @Component, @Autowired.
8. С помощью аннотации @PathVariable добавить возможность получать каждый фильм и данные о пользователях по их уникальному идентификатору: GET .../users/{id}.
9. Добавить методы, позволяющие пользователям добавлять друг друга в друзья, получать список общих друзей и лайкать фильмы. PUT /users/{id}/friends/{friendId} — добавление в друзья. DELETE /users/{id}/friends/{friendId} — удаление из друзей. GET /users/{id}/friends — возвращаем список пользователей, являющихся его друзьями. GET /users/{id}/friends/common/{otherId} — список друзей, общих с другим пользователем. PUT /films/{id}/like/{userId} — пользователь ставит лайк фильму. DELETE /films/{id}/like/{userId} — пользователь удаляет лайк. GET /films/popular?count={count} — возвращает список из первых count фильмов по количеству лайков. Если значение параметра count не задано, верните первые 10.
10. Добавить дополнительное логирование
11. Подключить к проекту дополнительный стартер: <dependency> <groupId>org.zalando</groupId> <artifactId>logbook-spring-boot-starter</artifactId> <version>3.7.2</version> </dependency>
12. Добавить в конфигурационный файл следующее свойство: logging.level.org.zalando.logbook: TRACE.

Akmula committed:
Версия 2.1

1. Создать интерфейсы FilmStorage и UserStorage.
2. Создать классы InMemoryFilmStorage и InMemoryUserStorage, имплементирующие новые интерфейсы, и перенести туда всю логику хранения, обновления и поиска объектов.
3. Добавить к InMemoryFilmStorage и InMemoryUserStorage аннотацию @Component, чтобы впоследствии пользоваться внедрением зависимостей и передавать хранилища сервисам.
4. Создать UserService, который будет отвечать за такие операции с пользователями, как добавление в друзья, удаление из друзей, вывод списка общих друзей.
5. Создать FilmService, который будет отвечать за операции с фильмами — добавление и удаление лайка, вывод 10 наиболее популярных фильмов по количеству лайков.
6. Переделатье код в контроллерах, сервисах и хранилищах под использование внедрения зависимостей.
7. Использовать аннотации @Service, @Component, @Autowired.
8. С помощью аннотации @PathVariable добавить возможность получать каждый фильм и данные о пользователях по их уникальному идентификатору: GET .../users/{id}.
9. Добавить методы, позволяющие пользователям добавлять друг друга в друзья, получать список общих друзей и лайкать фильмы.
	PUT /users/{id}/friends/{friendId} — добавление в друзья.
	DELETE /users/{id}/friends/{friendId} — удаление из друзей.
	GET /users/{id}/friends — возвращаем список пользователей, являющихся его друзьями.
	GET /users/{id}/friends/common/{otherId} — список друзей, общих с другим пользователем.
	PUT /films/{id}/like/{userId} — пользователь ставит лайк фильму.
	DELETE /films/{id}/like/{userId} — пользователь удаляет лайк.
	GET /films/popular?count={count} — возвращает список из первых count фильмов по количеству лайков. Если значение параметра count не задано, верните первые 10.
10. Добавить дополнительное логирование
11. Подключить к проекту дополнительный стартер:
	<dependency>
		<groupId>org.zalando</groupId>
		<artifactId>logbook-spring-boot-starter</artifactId>
		<version>3.7.2</version>
	</dependency>
12. Добавить в конфигурационный файл следующее свойство: logging.level.org.zalando.logbook: TRACE.

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/ErrorHandler.java:
```diff
@@ -0,0 +1,67 @@
+
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
+    public ErrorResponse handleInternalServerError(final RuntimeException e) {
```
public ErrorResponse handleThrowable(final Throwable e) {
стоит отлавливать все возможные исключения, поэтому стоит отлавливать Throwable

Akmula replied:
Изменил

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -2,67 +2,65 @@
+    @Autowired
+    private FilmStorage filmStorage;
+    @Autowired
+    private FilmService filmService;
```
final

YazykovMikhail replied:
попарвь везде

Akmula replied:
Поправил.

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -2,67 +2,65 @@
-    private final Map<Integer, Film> films = new HashMap<>();
+
+    @Autowired
+    private FilmStorage filmStorage;
```
Контроллер ходит только в сервис, а уже сервис ходить куда дальше нужно(другие сервиса или стораджи)
Каждый слой отвечает за свое, так что контроллер должен ходить в сервис, даже если в методе нет логики и она будет просто дублироваться. 
Это нужно, что бы при необходимости любой из слоев можно было заменить

Поправь и для User

Akmula replied:
Поправил

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -2,67 +2,65 @@
-import java.util.HashMap;
-import java.util.Map;
 
 @Slf4j
```
В методах контроллера нет логирвоания ответа
Поправь и для User

Akmula replied:
Сделано

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -7,22 +7,34 @@
     private Integer duration;
+    private Set<Integer> likes;
+
+    public void addLike(Integer id) {
```
Я предлагаю использовать отдельное поле для автоматического подсчета лайков:

	@JsonIgnore
	private long rate = 0;

	public void addLike(long userId) {
		likes.add(userId);
		rate = likes.size();
	}

	public void removeLike(long userId) {
		likes.remove(userId);
		rate = likes.size();
	}

Akmula replied:
Добавил.

Akmula replied:
@JsonIgnore
Если использовать эту аннотацию к private long rate = 0, то не проходят тесты постман.

YazykovMikhail replied:
Надо смотреть что не так, но обрати внимание что я подскзаал сделать:
private long rate = 0;

у тебя 
private Integer rate;

Это не проблема, можно и так, но возможно из за этого тесты и не проходили.

Akmula replied:
@NoArgsConstructor
@AllArgsConstructor
Добавил эти аннотации и все заработало.

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -7,22 +7,34 @@
 
     @Positive(message = "Продолжительность фильма должна быть положительным числом!")
     private Integer duration;
+    private Set<Integer> likes;
```
Используй @JsonIgnore что бы данные нельзя было поменять в методах контроллеров которые относяться для работы Film, ведь для лайков есть отедльные эндпоинты, а тут они просто храняться

Akmula replied:
Добавил

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -20,9 +21,17 @@ public class User {
 
     @Past(message = "Дата рождения не может быть в будущем!")
     private LocalDate birthday;
+    private Set<Integer> friends;
```
Используй @JsonIgnore что бы данные нельзя было поменять в методах контроллеров которые относяться для работы User, ведь для друзей есть отедльные эндпоинты, а тут они просто храняться

Akmula replied:
Добавил

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/service/IdGenerator.java:
```diff
@@ -0,0 +1,17 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class IdGenerator {
```
Это не очень хорошо. 
У тебя один сервис отвечает за разные данные, если что то ломается, то ломается сразу для всех.
Если нужно будет добавить новую сущность, нужно вносить сюда изменения. 
Большая связанность, это не очень. Лучше это убрать

Akmula replied:
Исправил

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -0,0 +1,82 @@
+public class InMemoryFilmStorage implements FilmStorage {
+    private static final int MAX_SIZE_DESCRIPTION = 200;
+    private static final LocalDate INTERNATIONAL_FILM_DAY = LocalDate.of(1895, 12, 28);
+    private final IdGenerator idGenerator;
```
Лучше сделать просто сделать полсе для генерации идентификаторов и все. 
Хранилище отвечет за генерацию идентификаторов - все ок

Akmula replied:
Исправил

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -0,0 +1,82 @@
+        return films.values();
+    }
+
+    private void validationFilm(Film film) {
```
Валидацию стоит вынести в сервис, в хранилище должны приходить данные уже полностью готовые для работы с хранилищем. 
Если они не валидные, им тут делать нечего, проверять бизнес данные - это дело сервиса

YazykovMikhail replied:
Аналогично для User

Akmula replied:
Сделано

Akmula committed:
Версия 2.2

1. Изменить ErrorResponse handleInternalServerError(final RuntimeException e) на ErrorResponse handleThrowable(final Throwable e).
2. Сделать финальными поля Storage и Service для контроллеров.
3. Удалить у контроллеров прямой доступ к хранилищам.
4. Добавить логирование ответа в методы контроллеров.
5. Добавить поле в Film для подсчета лайков.
6. Использовать @JsonIgnore для Set<Integer> likes и Set<Integer> friends.
7. Убрать @Service IdGenerator.
8. Убрать IdGenerator
9. Методы валидации перенести в сервисы

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,68 +1,79 @@
+        log.info("GET /films - Запрос на получение популярных фильмов. Выводить: {} фильмов", count);
+        Integer numberOfFilms = filmService.getPopular(count).size();
+        log.info("GET /films - Ответ на получение популярных фильмов. Вывод в кол-ве {} фильмов", numberOfFilms);
+        return ResponseEntity.ok(filmService.getPopular(count));
```
return ResponseEntity.ok(response));

YazykovMikhail replied:
Делать разные запросы для логов и самого ответа это не корректно, залогировал одно, а в ответ отправил другие. Ведь за это время могут поменять данные

YazykovMikhail replied:
Поправь везде, и в User тоже

Akmula replied:
Поправил

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,68 +1,79 @@
+    @GetMapping("/popular")
+    public ResponseEntity<Collection<Film>> getPopular(@RequestParam(defaultValue = "10") Integer count) {
+        log.info("GET /films - Запрос на получение популярных фильмов. Выводить: {} фильмов", count);
+        Integer numberOfFilms = filmService.getPopular(count).size();
```
Collection<Film> response = filmService.getPopular(count);

YazykovMikhail replied:
Лог с залогированным рахмером мало о чем говорит, размер может быть одинаковый, а данные разные

Akmula replied:
Исправил

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -1,75 +1,92 @@
-        return users.values();
+    public ResponseEntity<Collection<User>> getUsers() {
+        log.info("GET /users - Запрос на получение пользователей");
+        log.info("GET /users - Ответ на получение всех пользователей. Кол-во - {}", userService.getAllUsers().size());
```
как ты уже понял, это не очень

Akmula replied:
Исправлено

YazykovMikhail commented on src/main/java/ru/yandex/practicum/filmorate/exceptions/ErrorResponse.java:
```diff
@@ -0,0 +1,14 @@
+
+@Getter
+public class ErrorResponse {
+    String error;
```
Модификаторы доступа
 private

Akmula replied:
Исправлено

Akmula committed:
Версия 2.3

1. Сделать поля у class ErrorResponse приватными.
2. Изменить ответы в контроллерах и логах
